doc/TabletRouting.md

```diff
@@ -0,0 +1,194 @@
+  send the query to it. This was meant to handle cross-cell master fail-over.
+* For non-master queries, just route to the local cell vtgate is in.
+
+This turned out to be a bit limiting, as if no local tablet was available in the
+cell, the queries would just fail. We added the concept of Region, and if
+tablets in the same Region are available to serve, they can be used instead.
+This fix however is not entirely correct, as it uses the SrvKeyspace of the
+current cell, instead of the SrvKeyspace of the other cell. Both ServedFrom and
+the shard partition may be different in a different cell.
+
+## Percolating the cell back up
+
+We think the correct fix is to percolate the cell back up at the vtgate routing
+layer. That way when the higher level is asking for the SrvKeyspace object, it
+can ask for the right one, in the right cell. For instance, we can read the
+current SrvKeyspace for the current cell, see if it has healthy tablets for the
+shard(s) we're interested in, and if not find another cell in the same region
+that has healthy tablets.
+
+In the l2vtgate case though, this is not as easy, as the vtgate process has no
+idea what tablets are healthy. We propose to solve this by adding a healthcheck
+between vtgate and l2vtgate:
+
+* vtgate would maintain a healthcheck connection to the l2vtgate pool (a single
+  connection should be good enough, but we may be more aggressive to collect
+  more up to date information, so we don't depend on a single l2vtgate).
+* l2vtgate will report on a regular basis which keyspace / shard / tablet type
+  it can serve. It would also report the minimum and maximum replication lag
+  (and whatever other pertinent information vtgate needs).
+* l2vtgateGateway can then provide the health information back up inside vtgate.
+* This would also lift the limitation previously noted about l2vtgate going
+  cross-cell. When a master is moved from one cell to another, the l2vtgate in
+  the old cell would advertize back that they lost the ability to serve the
+  master tablet type, and the l2vtgate in the new cell would start advertizing
+  it. Vtgates would then know where to look.
+
+This would also be a good time to merge the vtgate code that uses the VSchema
+with the code that doesn't for SrvKeyspace access.
+
+## Hybrid Gateway
+
+It would be nice to re-organize the code a bit inside vtgate to allow for an
+hybrid gateway, and get rid of l2vtgate alltogether:
+
+* vtgate would use the discoveryGateway to watch the tablets in the current cell.
+* vtgate would use l2vtgateGateway to watch the tablets in a different cell.
+* vtgate would expose the RPC APIs currently exposed by the l2vtgate process.
+
+So vtgate would watch the tablets in the local cell only, but also know what
+healthy tablets are in the other cells, and be able to send query to them
+through their vtgate. The extra hop to the other cell vtgate should be a small
+latency price to pay, compared to going cross-cell already.
```

Comment:
I really like the cleanliness of this design, and in some environments this will likely be true, but for deployments like Slack is planning, I don't believe this is necessarily the case.

Specifically, AWS round trip times as measured from one of our vtgates is as follows:
```
from vtgate in us-east-1a:
tablet in us-east-1a: rtt min/avg/max/mdev = 0.100/0.117/0.133/0.012 ms
tablet in us-east-1b: rtt min/avg/max/mdev = 0.662/0.668/0.677/0.032 ms
tablet in us-east-1d: rtt min/avg/max/mdev = 0.401/0.408/0.419/0.007 ms
tablet in us-east-1e: rtt min/avg/max/mdev = 0.687/0.698/0.714/0.021 ms
```

So there is a clear penalty to cross AZs (both added latency and cost) but it is on the order of 4-6x, not orders of magnitude.

Still our plan is to deploy a vitess cell per AZ for both the latency benefits and fault tolerance, so I'd want to make sure this design still allows for the ability for one vtgate to cross cells.

More concretely, this design consideration is a consistency vs performance tradeoff -- Slack (and others) _might_ be willing to tolerate a small amount of time where queries could be misrouted during a shard split while the various SrvKeyspace records are being updated, if it means we don't have to pay an extra hop for every cross-cell query.

Reply:
I agree, and I think the proposed design would make this clear: you can still go cross-cell, but you need your vtgates to health-check the tablets cross-cell. That can be a large number of tablets. Otherwise just point at a vtgate pool somewhere else, and it will add the extra hop.

Reply:
Great! Thanks for clarifying.